### PR TITLE
regenesis: migrate FuelBlockIdsToHeights

### LINKED ISSUE #1790: Add migration for remaining off chain tables
@xgreenx wrote:

> We need to migrate as well:
> `FuelBlocks` - for that we need to disable guard on commit.
> `FuelBlockIdsToHeights` - Cab be generated form the `FuelBlocks`.
> `SpentMessages` - can be derived from Transactions or removed entirly
> `ProcessedTransactions` - can be derived from Transactions
> `FuelBlockConsensus`

and

> Actually, maybe transactions and blocks should be migrated to the off-chain database. Because we are only migrating them for GraphQL service and only to support queries for old data.

and

> [...] the idea is to have `OldFuelBlocks` and `OldTransactions` as part of the off-chain database. If the requested block is lower than genesis block height, we go to `off_chain::OldFuelBlocks` , otherwise to `on_chain::FuelBlocks`. In the case of transactions, yeah, first we go to on_chain::Transactions  and if no success, go to `off_chain::OldTransactions`

From the above, we should have a table for blocks and transactions in the off chain db and populate it upon regenesis.

### PR DESCRIPTION
**Pull request by Dentosal**
Closes #1790 by completing the last missing item.
Closes https://github.com/FuelLabs/fuel-core/issues/1545 by completing the last missing item as well.

This PR regenerates `FuelBlockIdsToHeights` for both blocks on regenesis.

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [ ] New behavior is reflected in tests

### Before requesting review
- [x] I have reviewed the code myself
- [ ] I have created follow-up issues caused by this PR and linked them here

### CODE REVIEW

File: crates/storage/src/transactional.rs [hunk 1]
```diff
@@ -190,13 +229,24 @@ where
...
     }
 }
 
 impl<Storage> Modifiable for InMemoryTransaction<Storage> {
-    fn commit_changes(&mut self, changes: Changes) -> StorageResult<()> {
+    /// Panics height policy mismatches the height policy fo the transaction.
+    fn commit_changes(
+        &mut self,
+        changes: Changes,
+        height_policy: ModifyHeightPolicy,
```
Comment (xgreenx) [inline crates/storage/src/transactional.rs:244]: Internal details of how the database works should not affect the interfaces.

**Review by xgreenx — COMMENTED** [review]
I think we need to introduce a `Genesis` phase into the database that allows violation of some rules. Or even create a new `GenesisDatabase` type that doesn't have guarantees/strict rules on the `commit_chagnes` method.

File: crates/storage/src/transactional.rs (hunk 1 above)
Comment (Dentosal) [reply crates/storage/src/transactional.rs]: I rewrote the PR to use a flag in the Database itself. The solution seems quite clean. Let me know what you think about this one.

File: crates/services/src/service.rs [hunk 2]
```diff
@@ -16,27 +18,6 @@ use std::any::Any;
 use tokio::sync::watch;
 use tracing::Instrument;
 
-/// Alias for `Arc<T>`
```
Comment (Dentosal) [inline crates/services/src/service.rs:19]: note: moved these to a separate module for clarity

File: crates/fuel-core/src/database.rs [hunk 3]
```diff
@@ -90,7 +90,10 @@ pub struct Database<Description = OnChain>
 where
     Description: DatabaseDescription,
 {
+    /// Cached value from Metadata table, used to speed up lookups.
```
Comment (Dentosal) [inline crates/fuel-core/src/database.rs:93]: See https://github.com/FuelLabs/fuel-core/pull/1694/commits/7b1141ac838568e3590f09dd420cb24a6946bd32#r1504911777 for the explanation behind this comment

File: crates/fuel-core/src/state.rs [hunk 4]
```diff
@@ -33,11 +33,7 @@ where
 
 pub trait TransactableStorage<Height>: IterableStore + Debug + Send + Sync {
     /// Commits the changes into the storage.
-    fn commit_changes(
-        &self,
-        height: Option<Height>,
-        changes: Changes,
-    ) -> StorageResult<()>;
+    fn commit_changes(&self, changes: Changes) -> StorageResult<()>;
```
Comment (xgreenx) [inline crates/fuel-core/src/state.rs:36]: We need `height: Option<Height>` for the state rewind feature=) I'm working on it right now, so I need this argument=D

I think maybe it makes sense to treat block height `None` as a regenesis and, at the end of the regenesis, set the block heights with information from the genesis block.

Comment (Dentosal) [reply crates/fuel-core/src/state.rs:36]: It would have been nice to leave a comment indicating that the useless argument was there for a reason.

Comment (Dentosal) [reply crates/fuel-core/src/state.rs]: Restored in 451f8be5d

**Review by xgreenx — APPROVED** [review]

**Pull request merged**